sw/host/opentitanlib/src/transport/ti50emulator/mod.rs

```diff
@@ -0,0 +1,183 @@
+                emulator: None,
+                spi_map: HashMap::new(),
+                gpio_map: HashMap::new(),
+                i2c_map: HashMap::new(),
+                uart_map: HashMap::new(),
+            })),
+        })
+    }
+}
+
+impl Drop for Ti50Emulator {
+    fn drop(&mut self) {
+        info!(
+            "Clenup Ti50Emulator instance directory: {}",
+            self.inner.borrow().instance_directory.display()
+        );
+        if let Err(e) = fs::remove_dir_all(&self.inner.borrow().instance_directory) {
+            error!("Can't remove instance directory error: {}", e)
+        }
+    }
+}
+
+/// Structure representing internal state of emulator
+struct Inner {
+    /// Path of parent directory representing instance.
+    instance_directory: PathBuf,
+    /// Path to a directory holding unmodified files (for example flash image).
+    _resource_directory: PathBuf,
+    /// Path to a directory containing working copies of image files and socket
+    /// used by the sub-process.
+    _runtime_directory: PathBuf,
+    /// Path to directory holding executable files - kernel, applications.
+    _executable_directory: PathBuf,
+    /// Current name of kernel executable files.
+    _executable: Option<String>,
+    /// `Emualtor` instance
+    emulator: Option<Rc<dyn Emulator>>,
+    /// HashMap for holding all instance of SPI indexed by its symbolic name.
+    spi_map: HashMap<String, Rc<dyn Target>>,
+    /// HashMap for holding all instance of GPIO pin indexed by its symbolic name.
+    gpio_map: HashMap<String, Rc<dyn GpioPin>>,
+    /// HashMap for holding all instance of I2C indexed by its symbolic name.
+    i2c_map: HashMap<String, Rc<dyn Bus>>,
+    /// HashMap for holding all instance of UART indexed by its symbolic name.
+    uart_map: HashMap<String, Rc<dyn Uart>>,
+}
+
+/// Implementation of the Transport trait backed based on TockOS HostEmulation port.
+impl Transport for Ti50Emulator {
+    fn capabilities(&self) -> Result<Capabilities> {
+        Ok(Capabilities::new(
+            Capability::UART
+                | Capability::GPIO
+                | Capability::SPI
+                | Capability::I2C
+                | Capability::EMULATOR,
+        ))
+    }
+
+    // Create SPI Target instance, or return one from a cache of previously created instances.
+    fn spi(&self, instance: &str) -> Result<Rc<dyn Target>> {
+        let inner = self.inner.borrow();
+        if let Some(spi) = inner.spi_map.get(instance) {
+            return Ok(Rc::clone(spi));
+        }
+        Err(
+            TransportError::InvalidInstance(TransportInterfaceType::Spi, instance.to_string())
+                .into(),
+        )
+    }
+
+    // Create I2C Target instance, or return one from a cache of previously created instances.
+    fn i2c(&self, instance: &str) -> Result<Rc<dyn Bus>> {
+        let inner = self.inner.borrow();
+        if let Some(i2c) = inner.i2c_map.get(instance) {
```

Comment:
A more idiomatic way to write this is

let i2c = self.inner.borrow().i2c_map.get().ok_or(TransportError::...())?;